Add Summary Collector

danvirsen opened the pull request:
This is basically [buffcode's code](https://github.com/buffcode/prometheus_client_php/commits/summary-collector) slightly tweaked to make it mergeable with this fork.

There are no tests and the only storage adapter that is implemented is Redis, so I wouldn't say it's ready for a merge just yet.

I'll try to get some time to take a look at tests and the other storage adapters, but I'm not sure when that might be. If someone else would like to take a look at it, they're more than welcome. 🙂

danvirsen committed:
Added summary collector (courtesy of buffcode)

danvirsen committed:
removed return type

NoelDavies commented on src/Prometheus/Summary.php:
```diff
@@ -88,13 +88,13 @@ public function observe($value, $labels = []): void
      * @return float
      */
-    public static function getQuantile($percentile, $values): float
+    public static function getQuantile($percentile, $values)
```
We should always include return types as we only support php 7.3+

NoelDavies commented on src/Prometheus/CollectorRegistry.php:
```diff
@@ -237,6 +242,66 @@ public function getOrRegisterHistogram($namespace, $name, $help, $labels = [], $
+     * @return Summary
+     * @throws MetricNotFoundException
+     */
+    public function getSummary($namespace, $name): Summary
```
function parameters should include strict type hints

NoelDavies reviewed (commented):
I've added a few comments, but so far it looks good. I think it'd be best to keep everything strictly typed to avoid any issues that may arise.

martinssipenko commented:
@danvirsen can you please update your branch to be up to date with master?

danvirsen committed:
added type hints

danvirsen committed:
phpcs formatting fixes

danvirsen committed:
added more type hints

danvirsen commented:
Sorry for the delay. I had email notifications turned off and didn't notice the responses.
I pulled in master, so it should be up to date.
I also added type hints and fixed some formatting issues that phpcs complained about.

martinssipenko commented:
@danvirsen tests are missing

danvirsen commented:
Yes, I'm aware. I will try to take some time to write tests, but I'm not sure when that will be.

NoelDavies commented on src/Prometheus/Storage/Redis.php:
```diff
@@ -348,6 +391,66 @@ private function collectHistograms(): array
+            $summaries[] = $summary;
+        }
+        return $summaries;
+    }
```
There are a LOT of nested loops in here, can we reduce these to improve performance?

NoelDavies commented on src/Prometheus/Summary.php:
```diff
@@ -0,0 +1,108 @@
+            $quantiles = self::getDefaultQuantiles();
+        }
+
+        if (0 === count($quantiles)) {
```
We need to flip these yoda statements to be consistent with the existing code base

marcospassos commented:
Any updates on this?

NoelDavies commented:
Hi @marcospassos, I'm afraid I no longer work at EndClothing. So one of the developers there should be maintaining this soon.

marcospassos commented:
@NoelDavies  Thank you for replying!
